File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(string dartResult) {
    int answer = 0;
    
    vector<int> V;
    
    for(int i=0; i<dartResult.size();i++){
        int tmp=0;
        
        if(isdigit(dartResult[i+1])){
            tmp = 10;
            i++;
        }
        
        else    tmp = dartResult[i]-'0';
        
        if(dartResult[i+1]=='S'){
            i++;
        }
                
        else if(dartResult[i+1]=='D'){
            tmp = pow(tmp,2);
            i++;
        }
                
        else if(dartResult[i+1]=='T'){
            tmp = pow(tmp,3);
            i++;
        }
            
            
        if(dartResult[i+1]=='*'){
            tmp *=2;
            if(!V.empty()){
                int before = V.back();
                V.pop_back();
                before*=2;
                V.push_back(before);
            }
            i++;
        }
        else if(dartResult[i+1]=='#'){
            tmp *= -1;
            i++;
        }

        
        V.push_back(tmp);
    }
    
    for(auto it:V){
        answer+=it;
    }
    
    
    return answer;
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
function solution(dartResult) {
    const arr = [];
    
    let number = '';
    
    for(let i=0; i<dartResult.length; i++){
        const select = dartResult[i];
        
        // 숫자인 경우
        if(select>='0' && select<='9')  number+=select;
        // 2) S 또는 D 또는 T 인 경우
        else if(select==='S' || select==='D' || select==='T'){
            switch(select){
                case 'S':
                    arr.push(Number(number));
                    break;
                case 'D':
                    arr.push(Math.pow(Number(number), 2));
                    break;
                case 'T':
                    arr.push(Math.pow(Number(number), 3));
                    break;
            }
            number = '';
        }
        // 3) * 또는 # 인 경우
        else {
            if(select==='*'){
                if(arr.length===1)  arr[arr.length-1] *= 2;
                else{
                    arr[arr.length-2] *= 2;
                    arr[arr.length-1] *= 2;
                }
            }
            else if(select==='#'){
                arr[arr.length-1] *= -1;
            }
        }
    }
    
    return arr.reduce((acc, cur)=> acc + cur, 0);
}